<font color="Lime"><h1 align ='center'><b><u>Vectores one-hot</u></b></h1></font>

Hasta ahora, hemos abordado la codificación de datos continuos (números de punto flotante), datos ordinales (normalmente enteros) y datos categóricos binarios, como supervivientes/fallecidos, hombres/mujeres.

Aquí veremos cómo codificar y los datos categóricos que tienen más de dos clases. También exploraremos cómo algunas decisiones que tomamos para mejorar nuestros modelos pueden en realidad afectar a su rendimiento.

## <font color='B8E139'><b>Los datos categóricos no son numéricos</b></font>

Los datos categóricos no son números de la misma manera que otros tipos de datos. Con los datos ordinales o continuos (numéricos), los valores más altos implican un aumento de la cantidad. Por ejemplo, en el Titanic, un precio de billete de 30 £ es más caro que el de un billete de 12 £.

Intentar codificar como números las características categóricas con más de dos clases genera problemas, ya que los datos categóricos no tienen un orden lógico.

Por ejemplo, Puerto de embarque tiene tres valores: C (Cherbourg), Q (Queenstown) y S (Southampton). No hay ninguna forma correcta de reemplazar estos símbolos por números, ya que hacerlo implicaría que uno de estos puertos es "menor que" el resto, mientras que otro sería "mayor que" los otros. No tiene sentido.

Como ejemplo de este problema, pasaremos a otra cosa y modelaremos la relación entre el Puerto de embarque y la Clase de billete, tratando el Puerto de embarque como un número. En primer lugar, se establece C < S < Q:

<p align ="center">
<img src = "https://raw.githubusercontent.com/JuanCamilo18/MM_Datos/main/imags/IDatos/Tdatos01.JPG">
</p>

En el trazado anterior, la línea predice un valor Pclass de ~3 para el Puerto Q.

Ahora, al establecer S < C< Q, se obtienen una predicción y una línea de tendencia completamente distintas:

<div>
<p align ="center">
<img src = "https://raw.githubusercontent.com/JuanCamilo18/MM_Datos/main/imags/IDatos/Tdatos01.JPG">
</p>
</div>

Ninguna de estas líneas de tendencia es correcta; no tiene sentido tratar las categorías como características continuas. Por lo tanto, ¿cómo se trabaja con las categorías?

## <font color='B8E139'><b>Codificación one-hot</b></font>

La codificación one-hot es una forma de codificar datos categóricos que evita el problema anterior. Cada categoría disponible obtiene su columna propia y una fila determinada únicamente contiene un solo 1 en la categoría a la que pertenece.

Por ejemplo, nuestro puerto se codificaría como tres columnas: una para Cherbourg, otra para Queenstown y otra para Southampton (el orden no es importante). Una persona que embarcara en Cherbourg tendría un 1 en la columna Port_Cherbourg, de esta forma:

<div align="center">

|Port_Cherbourg	| Port_Queenstown	| Port_Southampton|
|:-----:|:-----:|:-----:|
| 1	| 0	| 0|

</div>

Una persona que embarcara en Queenstown tendría un 1 en la segunda columna:

<div align="center">

|Port_Cherbourg	| Port_Queenstown	| Port_Southampton|
|:-----:|:-----:|:-----:|
| 0	| 1	| 0|

</div>

Una persona que embarcara en Southampton tendría un 1 en la tercera columna:

<div align="center">

|Port_Cherbourg	| Port_Queenstown	| Port_Southampton|
|:-----:|:-----:|:-----:|
| 0	| 0	| 1|

</div>

## <font color='B8E139'><b>Codificación one-hot, limpieza de datos y potencia estadística</b></font>

Antes de usar la codificación one-hot, es importante comprender que su uso puede tener un impacto positivo o negativo en el rendimiento real de un modelo.

### <font color='B8E139'><b>¿Qué es la potencia estadística?</b></font>

La potencia estadística hace referencia a la capacidad de un modelo para identificar de forma confiable las relaciones reales entre las características y las etiquetas. Por ejemplo, un modelo con potencia podría informar de la relación entre el precio del billete y la tasa de supervivencia con un alto grado de certeza. Por el contrario, un modelo con una potencia estadística baja podría no encontrar esta relación o notificarla con un bajo grado de certeza.

No entraremos en matemáticas, pero es importante tener en cuenta que determinadas opciones que tomamos pueden afectar a la potencia de nuestros modelos.

### <font color='B8E139'><b>La eliminación de datos reduce la potencia estadística</b></font>

Hemos hablado varias veces sobre cómo limpiar los datos mediante la eliminación de muestras que están incompletas. Un efecto colateral indeseado es que, al hacerlo, también se reduce la potencia estadística. Por ejemplo, supongamos que queremos predecir la supervivencia dados los datos siguientes:

<div align="center">

|Precio del billete	| Supervivencia |
|:-----:|:-----:|
|4 £	| 0|
|8 £ |	0|
|10 £ |	1|
|25 £ |	1|

</div>

A partir de estos datos, podríamos suponer que los pasajeros con un billete de 15 £ sobrevivieron, porque todas las personas con billetes con un precio superior a 10 £ sobrevivieron. Sin embargo, si tuviéramos menos datos, esto sería más difícil de suponer:

<div align="center">

|Precio del billete	| Supervivencia |
|:-----:|:-----:|
|4 £	| 0|
|8 £ |	0|
|25 £ |	1|

</div>

## <font color='B8E139'><b>Las columnas sin valor reducen la potencia estadística</b></font>

La potencia estadística también puede verse perjudicada al proporcionar modelos con características de poco valor, sobre todo cuando el número de características (columnas) comienza a aproximarse al número de muestras (filas).

Por ejemplo, supongamos que queremos predecir la supervivencia dados los datos siguientes:

<div align="center">

|Precio del billete	| Supervivencia |
|:-----:|:-----:|
|4 £	| 0|
|4 £	| 0|
|25 £ |	1|
|25 £ |	1|

</div>

Podríamos usar estos datos para predecir con confianza que un pasajero con un billete de 25 £ para la Cabina A sobreviviría, porque todos los pasajeros con billetes de 25 £ sobrevivieron.

Imagine, sin embargo, que tenemos otra característica (Cabina):

<div align="center">

|Precio del billete	| Cabin | Supervivencia |
|:-----:|:-----:|:-----:|
|4 £	| A | 0|
|4 £	| A | 0|
|25 £ |	B | 1|
|25 £ |	B | 1|

</div>

La columna Cabina no proporciona información útil, ya que simplemente corresponde al precio del billete. Ahora no queda claro si aquellos con un billete de 25 £ para la Cabina A sobrevivirían. ¿Fallecieron, como otros pasajeros de la Cabina A, o sobrevivieron como aquellos con billetes de 25 £?

## <font color='B8E139'><b>La codificación one-hot puede reducir la potencia estadística</b></font>

La codificación one-hot reduce la potencia estadística mucho más que los datos continuos u ordinales porque requiere varias columnas: una para cada valor categórico posible. Por ejemplo, la codificación one-hot del puerto de embarque agrega tres entradas de modelo (C, S y Q).

``
Como regla general, merece la pena incluir una variable categórica si el número de categorías es considerablemente menor que el número de muestras (filas del conjunto de datos) y esta categoría proporciona información que aún no está disponible para el modelo a través de otras entradas.
``

Por ejemplo, vimos que la probabilidad de supervivencia era distinta para las personas que embarcaron en puertos diferentes. Probablemente, esto solo refleja que la mayoría de las personas del puerto Queenstown tenían billetes de tercera clase. Por lo tanto, es probable que el embarque reduzca ligeramente la potencia estadística sin agregar información relevante a nuestro modelo.

Por el contrario, es probable que la Cabina influya considerablemente en la supervivencia, ya que las cabinas de la parte inferior del barco se llenarían de agua antes que las más cercanas a la cubierta. Dicho esto, el conjunto de datos del Titanic contiene 147 cabinas distintas, lo cual reducirá la potencia estadística del modelo si se incluye. Para averiguar si la Cabina es útil en definitiva, puede que tengamos que experimentar con su inclusión o exclusión en nuestro modelo.

En el ejercicio siguiente, por fin crearemos nuestro modelo de predicción de la supervivencia en el Titanic y practicaremos una codificación one-hot durante este proceso.



---



## <font color='B8E139'><b>Ejercicio: predicción de valores desconocidos con vectores one-hot</b></font>


## <font color='B8E139'><b>Ejercicio: Conjunto de datos Titanic - Vectores One-Hot</b></font>

En esta unidad, construiremos un modelo para predecir quién sobrevivió al desastre del Titanic. Practicaremos la transformación de datos entre tipos numéricos y categóricos, incluido el uso de vectores one-hot.

### <font color='B8E139'><b>Preparación de datos</b></font>

Primero, abriremos y limpiaremos rápidamente nuestro conjunto de datos, como hicimos en la última unidad:

In [1]:
import pandas
!wget https://raw.githubusercontent.com/MicrosoftDocs/mslearn-introduction-to-machine-learning/main/m0c_logistic_regression.py
!wget https://raw.githubusercontent.com/MicrosoftDocs/mslearn-introduction-to-machine-learning/main/Data/titanic.csv

# Open our dataset from file
dataset = pandas.read_csv("titanic.csv", index_col=False, sep=",", header=0)

# Fill missing cabin information with 'Unknown'
dataset["Cabin"].fillna("Unknown", inplace=True)

# Remove rows missing Age information
dataset.dropna(subset=["Age"], inplace=True)

# Remove the Name, PassengerId, and Ticket fields
# This is optional; it makes it easier to read our print-outs
dataset.drop(["PassengerId", "Name", "Ticket"], axis=1, inplace=True)

dataset.head()

--2023-03-29 18:15:19--  https://raw.githubusercontent.com/MicrosoftDocs/mslearn-introduction-to-machine-learning/main/m0c_logistic_regression.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 827 [text/plain]
Saving to: ‘m0c_logistic_regression.py’

m0c_logistic_regres 100%[===================>]     827  --.-KB/s    in 0s      

2023-03-29 18:15:19 (53.4 MB/s) - ‘m0c_logistic_regression.py’ saved [827/827]

--2023-03-29 18:15:19--  https://raw.githubusercontent.com/MicrosoftDocs/mslearn-introduction-to-machine-learning/main/Data/titanic.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... conne

Survived  Pclass     Sex   Age  SibSp  Parch     Fare    Cabin Embarked
0         0       3    male  22.0      1      0   7.2500  Unknown        S
1         1       1  female  38.0      1      0  71.2833      C85        C
2         1       3  female  26.0      0      0   7.9250  Unknown        S
3         1       1  female  35.0      1      0  53.1000     C123        S
4         0       3    male  35.0      0      0   8.0500  Unknown        S

## <font color='B8E139'><b>Acerca de nuestro modelo</b></font>

Entrenaremos un tipo de modelo conocido como Regresión logística, que predecirá quién sobrevive al desastre del Titanic.

Para este ejercicio, no necesita comprender la regresión logística. Colocamos la implementación fuera de este cuaderno en un método llamado `train_logistic_regression`. Si tiene curiosidad, puede leer este método en nuestro repositorio de GitHub.

`train_logistic_regression`:

1. Acepta nuestro marco de datos y una lista de características para incluir en el modelo.
2. Entrena al modelo.
3. Devuelve un número que indica el rendimiento del modelo al predecir la supervivencia de los pasajeros. `Los números más pequeños son mejores`.

### <font color='B8E139'><b>Solo numérico</b></font>

Vamos a crear un modelo que use solo las características numéricas.

Primero, lo usaremos `Pclass` aquí como una característica ordinal, en lugar de una característica categórica codificada en caliente.

In [2]:
from m0c_logistic_regression import train_logistic_regression

features = ["Age", "Pclass", "SibSp", "Parch", "Fare"] 
loss_numerical_only = train_logistic_regression(dataset, features)

print(f"Numerical-Only, Log-Loss (cost): {loss_numerical_only}")

Numerical-Only, Log-Loss (cost): 0.612168278948346


Tenemos nuestro punto de partida. Veamos si las características categóricas mejorarán el modelo.

## <font color='B8E139'><b>Características categóricas binarias</b></font>

Las características categóricas con solo dos valores potenciales se pueden codificar en una sola columna, como 0 o 1.

Convertiremos Sex valores en IsFemale- a 0 para hombre y 1 para mujer - e incluiremos eso en nuestro modelo.



In [3]:
# Swap male / female with numerical values
# We can do this because there are only two categories
dataset["IsFemale"] = dataset.Sex.replace({'male':0, 'female':1})

# Print out the first few rows of the dataset
print(dataset.head())

# Run and test the model, also using IsFemale this time
features = ["Age", "Pclass", "SibSp", "Parch", "Fare", "IsFemale"] 
loss_binary_categoricals = train_logistic_regression(dataset, features)

print(f"\nNumerical + Sex, Log-Loss (cost): {loss_binary_categoricals}")

   Survived  Pclass     Sex   Age  SibSp  Parch     Fare    Cabin Embarked  \
0         0       3    male  22.0      1      0   7.2500  Unknown        S   
1         1       1  female  38.0      1      0  71.2833      C85        C   
2         1       3  female  26.0      0      0   7.9250  Unknown        S   
3         1       1  female  35.0      1      0  53.1000     C123        S   
4         0       3    male  35.0      0      0   8.0500  Unknown        S   

   IsFemale  
0         0  
1         1  
2         1  
3         1  
4         0  

Numerical + Sex, Log-Loss (cost): 0.47071447420412404



¡Nuestra pérdida (error) disminuyó! Este modelo funciona mejor que el modelo anterior.

## <font color='B8E139'><b>Codificación One-Hot</b></font>

La clase de ticket ( Pclass) es una característica ordinal. Sus valores potenciales (1, 2 y 3) tienen un orden y tienen el mismo espaciamiento. Sin embargo, este espaciado uniforme podría ser incorrecto: en las historias sobre el Titanic, los pasajeros de tercera clase fueron tratados mucho peor que los de primera y segunda clase.

Convirtámoslo Pclass en una característica categórica usando una codificación one-hot:

In [4]:
# Get all possible categories for the "PClass" column
print(f"Possible values for PClass: {dataset['Pclass'].unique()}")

# Use Pandas to One-Hot encode the PClass category
dataset_with_one_hot = pandas.get_dummies(dataset, columns=["Pclass"], drop_first=False)

# Add back in the old Pclass column, for learning purposes
dataset_with_one_hot["Pclass"] = dataset.Pclass

# Print out the first few rows
dataset_with_one_hot.head()

Possible values for PClass: [3 1 2]


Survived     Sex   Age  SibSp  Parch     Fare    Cabin Embarked  IsFemale  \
0         0    male  22.0      1      0   7.2500  Unknown        S         0   
1         1  female  38.0      1      0  71.2833      C85        C         1   
2         1  female  26.0      0      0   7.9250  Unknown        S         1   
3         1  female  35.0      1      0  53.1000     C123        S         1   
4         0    male  35.0      0      0   8.0500  Unknown        S         0   

   Pclass_1  Pclass_2  Pclass_3  Pclass  
0         0         0         1       3  
1         1         0         0       1  
2         0         0         1       3  
3         1         0         0       1  
4         0         0         1       3


Tenga en cuenta que Pclassse convirtió en tres valores: Pclass_1, Pclass_2y Pclass_3.

Las filas con Pclass valores de 1 tienen un valor en la Pclass_1 columna. Vemos el mismo patrón para los valores de 2 y 3.

Ahora, vuelva a ejecutar el modelo y trate Pclass los valores como valores categóricos, en lugar de valores ordinales.

In [5]:
# Run and test the model, also using Pclass as a categorical feature this time
features = ["Age", "SibSp", "Parch", "Fare", "IsFemale",
            "Pclass_1", "Pclass_2", "Pclass_3"]

loss_pclass_categorical = train_logistic_regression(dataset_with_one_hot, features)

print(f"\nNumerical, Sex, Categorical Pclass, Log-Loss (cost): {loss_pclass_categorical}")


Numerical, Sex, Categorical Pclass, Log-Loss (cost): 0.47171118251126615



¡Esto parece haber empeorado un poco las cosas!

Vamonos.

## <font color='B8E139'><b>Incluye cabina</b></font>

Recordemos que muchos pasajeros tenían Cabin información. Cabin es una característica categórica y debería ser un buen predictor de supervivencia, porque las personas en las cabinas inferiores probablemente tuvieron poco tiempo para escapar durante el hundimiento.

Codifiquemos la cabina usando vectores one-hot e incluyémosla en un modelo. Esta vez, hay tantas cabañas que no las imprimiremos todas. Para practicar imprimiéndolos, siéntete libre de editar el código como práctica.

In [6]:
# Use Pandas to One-Hot encode the Cabin and Pclass categories
dataset_with_one_hot = pandas.get_dummies(dataset, columns=["Pclass", "Cabin"], drop_first=False)

# Find cabin column names
cabin_column_names = list(c for c in dataset_with_one_hot.columns if c.startswith("Cabin_"))

# Print out how many cabins there were
print(len(cabin_column_names), "cabins found")

# Make a list of features
features = ["Age", "SibSp", "Parch", "Fare", "IsFemale",
            "Pclass_1", "Pclass_2", "Pclass_3"] + \
            cabin_column_names

# Run the model and print the result
loss_cabin_categorical = train_logistic_regression(dataset_with_one_hot, features)

print(f"\nNumerical, Sex, Categorical Pclass, Cabin, Log-Loss (cost): {loss_cabin_categorical}")

135 cabins found

Numerical, Sex, Categorical Pclass, Cabin, Log-Loss (cost): 0.460030454816552



¡Ese es nuestro mejor resultado hasta ahora!

### <font color='B8E139'><b>Mejorar el poder</b></font>

Incluir un gran número de clases categóricas, por ejemplo, 135 cabinas, a menudo no es la mejor manera de entrenar un modelo, porque el modelo solo tiene algunos ejemplos de cada clase de categoría para aprender.

A veces, podemos mejorar los modelos si simplificamos las características. Cabin probablemente fue útil porque indicaba en qué cubierta del Titanic probablemente se encontraban las personas: aquellos en las cubiertas inferiores habrían inundado sus habitaciones primero.

Podría resultar más sencillo utilizar la información de cubierta, en lugar de categorizar a las personas en camarotes.

Simplifiquemos lo que hemos ejecutado, reemplazando las 135 `Cabin` categorías con una `Deck` categoría más simple que tiene solo 9 valores: A - G, T y U (Desconocido)

In [7]:
# We have cabin names, like A31, G45. The letter refers to the deck that
# the cabin was on. Extract just the deck and save it to a column. 
dataset["Deck"] = [c[0] for c in dataset.Cabin]

print("Decks: ", sorted(dataset.Deck.unique()))

# Create one-hot vectors for:
# Pclass - the class of ticket. (This could be treated as ordinal or categorical)
# Deck - the deck that the cabin was on
dataset_with_one_hot = pandas.get_dummies(dataset, columns=["Pclass", "Deck"], drop_first=False)

# Find the deck names
deck_of_cabin_column_names = list(c for c in dataset_with_one_hot.columns if c.startswith("Deck_"))
 
features = ["Age", "IsFemale", "SibSp", "Parch", "Fare", 
            "Pclass_1", "Pclass_2", "Pclass_3",
            "Deck_A", "Deck_B", "Deck_C", "Deck_D", 
            "Deck_E", "Deck_F", "Deck_G", "Deck_U", "Deck_T"]

loss_deck = train_logistic_regression(dataset_with_one_hot, features)

print(f"\nSimplifying Cabin Into Deck, Log-Loss (cost): {loss_deck}")

Decks:  ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'T', 'U']

Simplifying Cabin Into Deck, Log-Loss (cost): 0.45886536352547463


### <font color='B8E139'><b>Comparación de modelos</b></font>

Comparemos los `loss` de estos modelos:

In [8]:
# Use a dataframe to create a comparison table of metrics
# Copy metrics from previous Unit

l =[["Numeric Features Only", loss_numerical_only],
    ["Adding Sex as Binary", loss_binary_categoricals],
    ["Treating Pclass as Categorical", loss_pclass_categorical],
    ["Using Cabin as Categorical", loss_cabin_categorical],
    ["Using Deck rather than Cabin", loss_deck]]

pandas.DataFrame(l, columns=["Dataset", "Log-Loss (Low is better)"])

Dataset  Log-Loss (Low is better)
0           Numeric Features Only                  0.612168
1            Adding Sex as Binary                  0.470714
2  Treating Pclass as Categorical                  0.471711
3      Using Cabin as Categorical                  0.460030
4    Using Deck rather than Cabin                  0.458865


Podemos ver que incluir características categóricas puede mejorar y perjudicar el funcionamiento de un modelo. A menudo, la experimentación es la mejor manera de encontrar el mejor modelo.

**Resumen**

En esta unidad, aprendió a usar la codificación One-Hot para abordar datos categóricos.

También exploramos cómo, a veces, el pensamiento crítico sobre el problema en cuestión puede mejorar una solución más que simplemente incluir todas las características posibles en un modelo.